features/netsocket/TCPSocket.cpp

```diff
@@ -28,17 +29,26 @@ TCPSocket::TCPSocket(TCPSocket *parent, nsapi_socket_t socket, SocketAddress add
     _stack = parent->_stack;
     _factory_allocated = true;
     _remote_peer = address;
-
+    _socket_stats.stats_new_socket_entry(this);
     _event = mbed::Callback<void()>(this, &TCPSocket::event);
     _stack->socket_attach(socket, &mbed::Callback<void()>::thunk, &_event);
 }
 
 TCPSocket::~TCPSocket()
 {
+    _socket_stats.stats_update_socket_state(this, SOCK_CLOSED);
+}
+
+nsapi_error_t TCPSocket::close()
+{
+    _socket_stats.stats_update_socket_state(this, SOCK_CLOSED);
+    return InternetSocket::close();
 }
 
 nsapi_protocol_t TCPSocket::get_proto()
 {
+    _socket_stats.stats_update_proto(this, NSAPI_TCP);
```

Comment:
This is not in the correct place.

` _socket_stats.stats_update_proto(this, NSAPI_TCP);` should be in the contructor.
You cannot change the socket type after creating it.